### DISCUSSION TITLE
ByLaw founder organisation

### DISCUSSION THREAD

**Discussion by glennawatson** [dotnet-foundation/Home#49] [General Discussions]
There's been some discussion asking if the foundation bylaws having specific reference to Microsoft as the founding organisation and special veto rights creates a conflict of interest.

Is this even on the table to change at the moment? Is it important?

When they think of .NET, people generally associate Microsoft as the most significant contributor to the source code.

**Comment by Pilchard123** [comment]
> Microsoft as the founding organisation

I think I first learned about the .NET Foundation during .NET Conf 2020. I remember thinking at the time "this sounds very much like Microsoft pretending it is not Microsoft". I did not know about the Founding Member status or the director-with-veto, but now I think it sounds even more like "Microsoft pretending it is not Microsoft".

>When they think of .NET, people generally associate Microsoft as the most significant contributor to the source code.

If you were to say to me "quick, what is .NET in one sentence", I would probably reply something like this: "a runtime, a set of libraries, and tooling, chiefly developed by Microsoft". Now, I may be reading more into "when they think of .NET" than you intend, but I suspect that by ".NET" you mean "the runtime, libraries, tooling *as well as* the community and projects built around those things". I would disagree (though I can get on board with some poetic licence along the lines of _E Pluribus Unum_, but only poetically) and say that "the community and projects built around those things" is distinct from ".NET".

Which that in mind, when I think of .NET as

* a runtime/libraries/tooling
    * I generally associate Microsoft as the most significant contributor to the source code. If I thought for a while I might wonder if the open-source bits of .NET (mostly the non-Framework stuff) are more than 50% Microsoft-produced or if Microsoft is just the largest single contributor but in the minority overall.
*  communities and projects built around those things
    * I **do not** generally associate Microsoft as the most significant contributor to the source code. Individual projects may have never interacted with Microsoft further than using C# as the language that the project is developed in. To take my own projects as an example, none of my code has ever been touched by anyone from Microsoft. I might develop it using Microsoft-produced tools and using Microsoft-produced libraries (among others), but the project itself is nothing to do with Microsoft. If I thought Java or C++ or Malbolge was a better fit for the project I would have used that instead. To take another example, a couple of years ago I started playing around with IdentityServer. I did not think "this is part of the .NET Foundation, it must be significantly contributed to by Microsoft" - I did not even know the .NET Foundation existed! Nor did I think "this project uses .NET, it must be significantly contributed to by Microsoft". I did, however, think "_Dominick Baier and Brock Allen_ have made a good product".
 * in the context of closed-source/proprietary projects using .NET and their communities
    * I **do not** generally associate Microsoft as the most significant contributor to the source code.

**Reply by Pilchard123** [reply]
To be clear, I do think that the foundation (or something like it) is a good idea at its core, and I don't necessarily have any problem with it being Microsoft-but-not-Microsoft if that's made clear up-front. Many of the benefits listed on the foundation website are exactly the sort of thing I'd be looking for help with if I had a project that warranted it, and who better to know about Microsoft tools than Microsoft?

**Reply by glennawatson** [reply]
Yeah. There has been attempts to separate the foundation from Microsoft in the past by the board etc but there could be arguments for embracing it

**Reply by rprouse** [reply]
I've only been on the board for a week but so far I have felt no pressure or influence from Microsoft. There are Microsoft employees on the board but all of my interactions with them are as respected members of the .NET community not Microsoft employees. My experience so far is that Microsoft is there to support us but isn't exerting any control.

Let's not forget that .NET and nearly everything around it is open source and part of the foundation. Most of the people working on it are Microsoft employees, so it is only natural that they are represented. I can see the concern with the wording of the bylaws, I read it that way at first, but my experience so far has reduced my concerns.

**Reply by dggmez** [reply]
>I can see the concern with the wording of the bylaws, I read it that way at first, but my experience so far has reduced my concerns.

But that's the problem. The concerns were only reduced once you got on the board. That's really problematic because most of the members are not there. The only thing they have is the bylaws.

Now I get what @Pilchard123 and @rprouse are saying about how important Microsoft is for .NET. I can even buy that they did and do so much work that they *deserve* to have one director on the board like the current bylaws say. But such director having special veto rights and requiring its presence as part of any quorum (section 3.9 (e))? Requiring amendments that alter Microsoft's special rights at the foundation to be authorized by Microsoft itself (Article IX)? That's too much power for one member. I can't buy that. Or at least I can't buy that one member has so much power and at the same time that the foundation belongs to all the members.

**Reply by Perksey** [reply]
Regardless of whether Microsoft has exercised these rights or not, Microsoft having an explicit callout in the bylaws removes all plausible deniability of Microsoft doing so.

Logically speaking, if they're not exercising these rights particularly at the moment, removing them from the bylaws shouldn't make too much of a difference, but I appreciate in any case Microsoft will raise an eyebrow.

Removing Microsoft from the bylaws I feel is in the foundation's best interest, especially given the bad press at the moment... If Microsoft wants to make sure that their voice is still heard, perhaps:
1. reword the bylaws to ensure significant contributors are recognised, rather than explicitly calling out Microsoft
    - might not be the best as this creates a "leaderboard" for contributors wherein those at the top get special rights
2. reword the bylaws to ensure entities/corporate sponsors with a vested interest in .NET are recognised
    - again probably not the best, as it means anyone can just buy their way into the foundation which is a recipe for corruption
3. remove them from the bylaws and form another committee for corporations with a vested interest so that these entities are still able to advise the board, but not necessarily force their hand
    - this is probably my preferred option. still has the corporations such as Microsoft involved and able to influence the board, but they don't have "god rights" if you'd like.

**Reply by Pilchard123** [reply]
If I've understood right, I'm with @Perksey - Microsoft being given some special privilege leaves a nasty taste in the mouth at best, and I might even go further and say that it could be actively dangerous. The Microsoft of today might not be using those privileges (I do believe you, @rprouse, when you say you've felt no pressure or influence), and they might not have any plans to - but what about the Microsoft of five years hence?

**Reply by glennawatson** [reply]
At the moment microsoft does exercise their right in terms of having a assigned board member, haven't heard of them using the veto power though.

**Reply by Pilchard123** [reply]
I was more thinking of the veto power of the appointed director than the director themselves. I'd sort of treated the appointment as a given (at least, in the context of the current situation) since the bylaws require it but the veto power is discretionary.

**Reply by Perksey** [reply]
It's worth noting that removing the Founding Member "god rights" will also need sign-off from a Microsoft signatory above the appointed director (wouldn't be surprised if it has to go all the way up the SLT tbh) so any changes to the bylaws on this regard will have to be very carefully presented to Microsoft to ensure they don't feel like they're losing control over the .NET Platform (and if this is the case, we probably need to bake in stuff like what rights the foundation can exercise on member projects into the bylaws)

**Reply by robmen** [reply]
This comment prompted a very interesting thought (to me):

>  presented to Microsoft to ensure they don't feel like they're losing control over the .NET Platform

Let me replace the names with a similar (but much smaller) project in the .NET Foundation:

>  presented to FireGiant to ensure they don't feel like they're losing control over the WiX Toolset

Pick any project and replace the names. I wonder if that mad-lib provides a very simple way to frame the problem. We'd like the same sort of protections in the .NET Foundation that Microsoft enjoys. Right?

**Comment by gortok** [comment]
As I've [written](https://github.com/dotnet-foundation/Home/discussions/40#discussioncomment-1450074), and [written](https://twitter.com/gortok/status/1268536880427040769?s=20), and [spoken about elsewhere](https://podcast.lastweekin.net/4),

the issue is not that Microsoft has a seat on the board; the issue is that the .NET Foundation -- for all intents and purposes -- mimics an extension of Microsoft's desires for .NET.

How does this manifest itself?

1. Legal agreements that are not explained to the community; nor are they public (the contracts between maintainers and the foundation), and the questions asked about those contracts and [changes to how the foundation works are never answered](https://github.com/dotnet-foundation/projects/issues/122).  The legal agreements seem to greatly benefit the .NET foundation and microsoft, going so far as to replacing the copyright header for every project, regardless of whether they assign the copyright to the .NET Foundation or not.

2. The Executive Director is paid by Microsoft. The Treasurer is paid by Microsoft. Microsoft gets a special seat on the board that is not only irrevocable, but keeps the structure of the .NET Foundation from changing with a special veto.  This would be one thing if they were not a Microsoft employee -- but how many non Microsoft Employees have been the Executive Director of the foundation? Even this wouldn't be an issue if there had ever been a feeling that the Executive Director was independent in thought and deed from the Microsoft legal entity; but I haven't seen that yet, out of any of the Executive Directors (special note: I think they're all Good People, and I am grateful for their service; but that doesn't change the outcome).

3. Instances where "what is good for the .NET community" and "what is in Microsoft's best interests" have -- to a T -- gotten resolved in the 'This is in Microsoft's best interests'.  The aforementioned model changes and subsequent 'just put our name in the copyright header', the 'commercially viable' changes to the mission statement (Which was never explained -- commercially viable *for whom*), heck, [even members of the .NET team see the issues](https://github.com/microsoft/dotnet/blob/master/docs/ecosystem-issues.md). The tone-deafness of the Maturity Model comes to mind as well -- there's an institutional context in the .NET Foundation that seemingly represents the world from Microsoft's point of view, with little focus on the issues that actually affect the members.

Here's what an independent .NET foundation would look like:

1. A mission and vision statement that doesn't reflect Microsoft's view of .NET, but represents the idea that .NET, as its own economic ecosystem, has a set of goals and ideals that it looks towards, regardless of who pays the bills.  While I don't have words for what it is, I would tend to see the following ideas captured:

  - grow the ecosystem economically for the maintainers and community
  - reduce reliance on first-party solutions 
  - increase confidence and reliance on third-party solutions
  - provide a monetization strategy and legal assistance for projects that wish to monetize
  - be an interest group for the .NET community, and not stay silent on matters regarding the .NET ecosystem [even when the patron does something stupid](https://twitter.com/gortok/status/1268536880427040769?s=20).

2. A board who represents the community and who are vocal about their support for the .NET community.  Right now Board members say very little publicly; and we've had entire boards come and go *with no public discussions from board members*.  This is an unconscionable state of affairs.

3. A .NET Foundation that feels vibrant and full of action, and doesn't make us ask "What did they even do this past year?"

4. A plan and pathway to the ED and the salaried members of the .NET Foundation being independent from Microsoft.

This is what the .NET community needs from my perspective; I'd love to hear everyone's thoughts on this. This affects all of us.

**Reply by Perksey** [reply]
> Right now Board members say very little publicly

For the record I am very happy with the public discussion from the new board members, it's the most hands-on with the community a .NET Foundation board has been for as long as I can remember <3
